src/v/kafka/server/group_manager.cc

```diff
@@ -115,28 +116,25 @@ ss::future<> group_manager::stop() {
 
 void group_manager::detach_partition(const model::ntp& ntp) {
     klog.debug("detaching group metadata partition {}", ntp);
-    std::vector<group_ptr> groups;
-    groups.reserve(_groups.size());
-    for (auto& group : _groups) {
-        groups.push_back(group.second);
-    }
-
-    for (auto& gr : groups) {
-        // skip if group is not managed by current NTP
-        if (gr->partition()->ntp() != ntp) {
-            continue;
+    ssx::spawn_with_gate(_gate, [this, ntp]() -> ss::future<> {
+        auto it = _partitions.find(ntp);
+        if (it == _partitions.end()) {
+            co_return;
         }
-        auto it = _groups.find(gr->id());
-        if (it == _groups.end()) {
-            continue;
+        auto p = it->second;
+        auto units = co_await p->catchup_lock.hold_write_lock();
+
+        for (auto g_it = _groups.begin(); g_it != _groups.end();) {
+            if (g_it->second->partition()->ntp() == p->partition->ntp()) {
+                g_it->second->shutdown();
+                _groups.erase(g_it++);
+                continue;
+            }
+            ++g_it;
         }
-
-        vlog(klog.trace, "Removed group {}", gr);
-        _groups.erase(it);
-        _groups.rehash(0);
-    }
-    _partitions.erase(ntp);
-    _partitions.rehash(0);
+        _partitions.erase(it);
```

Comment:
this change looks like a segfault waiting to happen. what makes the iterator remain stable across a scheduling point?

```c++
auto it = _partitions.find(ntp);
auto units = co_await p->catchup_lock.hold_write_lock();
_partitions.erase(it);
```